components/net_traits/blob_url_store.rs

```diff
@@ -33,13 +34,25 @@ pub struct BlobBuf {
 
 /// Parse URL as Blob URL scheme's definition
 /// https://w3c.github.io/FileAPI/#DefinitionOfScheme
-pub fn parse_blob_url(url: &Url) -> Option<(Uuid, Option<&str>)> {
-    url.path_segments().and_then(|mut segments| {
-        let id_str = match (segments.next(), segments.next()) {
-            (Some(s), None) => s,
-            _ => return None,
-        };
+pub fn parse_blob_url(url: &Url) -> Result<(Uuid, FileOrigin, Option<String>), ()> {
+    let url_inner = try!(Url::parse(url.path()).map_err(|_| ()));
+    let fragment = url_inner.fragment().map(|s| s.to_string());
+    let mut segs = try!(url_inner.path_segments().ok_or(()));
+    let id = try!(segs.nth(0).ok_or(()));
+    let id = try!(Uuid::from_str(id).map_err(|_| ()));
+    Ok((id, get_blob_origin(&url_inner), fragment))
+}
 
-        Uuid::from_str(id_str).map(|id| (id, url.fragment())).ok()
-    })
+/// Given an URL, returning the Origin that a Blob created under this
+/// URL should have.
+/// HACK(izgzhen): Not well-specified on spec, and it is a bit a hack
+/// both due to ambiguity of spec and that we have to serialization the
+/// Origin here.
+pub fn get_blob_origin(url: &Url) -> FileOrigin {
+    if url.scheme() == "file" {
+        // NOTE: by default this is "null" (Opaque), which is not ideal
+        "file://".to_string()
```

Comment:
Aren't file URIs supposed to be considered opaque origin? What's wrong with using the opaque origin in the blob?